### ISSUE TITLE
Expose {slug:} in the Admin for all objects so that it can be updated to something human readable

### ISSUE THREAD

**Issue by willphdavies** [Solid-Interactive/grasshopper-admin#43]
Currently our code makes reference to mongo IDs or meta.labels in order to correctly filter content

This is time consuming and not great practice as IDs are not human readable and meta.labels are neither unique nor immutable - If I change a content type label in Cambium or Gerber it will break the site going forward in a way that is difficult to debug

I added functionality for .getBySlug to nodes and contenttypes as part of GH Ecomm but did not change the admin to make use of this.

Once we get this in place we will be closer to being able to define contenttypes in the codebase rather than in the database which is almost essential for creating plugins

**Comment by travism** [comment]
Would the slug change automatically when content is moved from 1 node to another?

**Comment by pajtai** [comment]
Don't think so. We're viewing slug as a unique identifier equivalent to _id, but just human readable. Trying to push two docs with the same slug to the same collection throws an error.

As an example, moving a piece of content with the slug "faq" from the "pages" node to the "about" node would keep the content's slug as "faq".

**Comment by willphdavies** [comment]
The idea is that every item in a collection would have a unique identifier
that can be defined by the developer in order to establish relationships.
If a piece of content is moved from one node to another then, if the node
slug is stored in that content item, it would have to be updated.

Really, I'd like to be able to express relationships in much the same way
as you do in keystone Js and i thought slugs might be a way to do that.
Maybe it's worth taking a look at the keystone Js source code to see what
they did

**Comment by pajtai** [comment]
When you move a piece of content from one node to the other, of course the meta.node field would be updated. I'm just saying the content's `slug` field should ideally never be updated, just like it's `_id` should ideally never be updated.

**Comment by willphdavies** [comment]
Agreed. The goal is to be able to define relationships without knowing the
ID. An immutable slug seems like the fastest route given that they already
exist

**Comment by travism** [comment]
It really seems like you guys are talking about changing the ID of the content. Creating relationships is typically done using an ID (indexed key - faster). Is there a benefit / detriment of doing this is mongo?

**Comment by travism** [comment]
Oh, additional thought. Sometimes content is created by users or dynamically. There might need to be an "auto-creation" of this `slug/id`

**Comment by pajtai** [comment]
The reason we were thinking of using slug vs id is that it'd be technically easier. Slug will be indexed too.

The reason slug is easier to deal with than _id is that there are multiple places in the core codebase where there is a regex which relies on _id to be of the form `^[0-9a-fA-F]{24}` (in content, node, getDeep).... This would be difficult to work around currently. It also is very convenient to be able to rely on the form of _id.

**Comment by willphdavies** [comment]
i also ran into problems based on the fact that you need to know a node ID
in order to hit the Add Content functionality on GH.

EG: Content Type - Page
Node - Pages (Accepts Content Types - Pages)

To add a new page :
/admin/items/nodeid/{NODEID}/create-content

This seems a very opaque way to get to URL add content to a collection
which is hard to achieve outside of the context of the ADMIN codebase (in a
plugin for instance).

A URL for adding content to a content type ideally would be:

/admin/items/add/{content-type}

Assigning a piece of content to a node seems like it would be better as a
separate operation in order to remove the dependency of one from the other
